drivers/digital-io/max22196/max22196.c

```diff
@@ -0,0 +1,428 @@
+ * @param cnt_msb_byte - Counter value to be set for the MSB Byte.
+ * @param cnt_lsb_byte - Counter value to be set for the LSB Byte.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_set_chan_cnt(struct max22196_desc *desc, uint32_t ch,
+			  uint8_t cnt_msb_byte, uint8_t cnt_lsb_byte)
+{
+	int ret;
+
+	if (ch > MAX22196_CHANNELS - 1)
+		return -EINVAL;
+
+	ret = max22196_reg_update(desc, MAX22196_START_STOP_REG,
+				  MAX22196_CNT_MASK(ch),
+				  no_os_field_prep(MAX22196_CNT_MASK(ch), 0));
+	if (ret)
+		return ret;
+
+	ret = max22196_reg_write(desc, MAX22196_CNT_LSB_REG(ch), cnt_lsb_byte);
+	if (ret)
+		return ret;
+
+	ret = max22196_reg_write(desc, MAX22196_CNT_MSB_REG(ch), cnt_msb_byte);
+	if (ret)
+		return ret;
+
+	return max22196_reg_update(desc, MAX22196_START_STOP_REG,
+				   MAX22196_CNT_MASK(ch),
+				   no_os_field_prep(MAX22196_CNT_MASK(ch), 1));
+}
+
+/**
+ * @brief MAX22196 channel counter get function
+ * @param desc - MAX22196 device descriptor.
+ * @param ch - Channel for which the channel counter is read.
+ * @param cnt_msb_byte - Variable through which the counter's msb byte is read.
+ * @param cnt_lsb_byte - Variable through which the counter's lsb byte is read.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_get_chan_cnt(struct max22196_desc *desc, uint32_t ch,
+			  uint8_t *cnt_msb_byte, uint8_t *cnt_lsb_byte)
+{
+	int ret;
+
+	if (ch > MAX22196_CHANNELS - 1)
+		return -EINVAL;
+
+	ret = max22196_reg_update(desc, MAX22196_START_STOP_REG,
+				  MAX22196_CNT_MASK(ch),
+				  no_os_field_prep(MAX22196_CNT_MASK(ch), 0));
+	if (ret)
+		return ret;
+
+	ret = max22196_reg_read(desc, MAX22196_CNT_LSB_REG(ch), cnt_lsb_byte);
+	if (ret)
+		return ret;
+
+	ret = max22196_reg_read(desc, MAX22196_CNT_MSB_REG(ch), cnt_msb_byte);
+	if (ret)
+		return ret;
+
+	return max22196_reg_update(desc, MAX22196_START_STOP_REG,
+				   MAX22196_CNT_MASK(ch),
+				   no_os_field_prep(MAX22196_CNT_MASK(ch), 1));
+}
+
+/**
+ * @brief Initialize and configure the MAX22196 device.
+ * @param desc - device descriptor for the MAX22196 that will be initialized.
+ * @param param - initialization parameter for the device.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int max22196_init(struct max22196_desc **desc,
+		  struct max22196_init_param *param)
+{
+	struct max22196_desc *descriptor;
+	uint32_t reg_val;
+	int ret;
+	int i;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+	ret = no_os_spi_init(&descriptor->comm_desc, param->comm_param);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_get_optional(&descriptor->crc_desc, param->crc_param);
+	if (ret)
+		goto spi_error;
+
+	if (descriptor->crc_desc) {
+		ret = no_os_gpio_get_value(descriptor->crc_desc, &reg_val);
+		if (ret)
+			goto gpio_error;
+		if (reg_val == NO_OS_GPIO_HIGH)
+			descriptor->crc_en = true;
+		else
+			descriptor->crc_en = false;
+	}
```

Comment:
This may also be set by the target microcontroller (as in the CRC pin being output), so the logic won't work in such case.

Reply:
I will change the code so CRC GPIO needs to be set as an output from the microcontroller (since it's an input in the case of the MAX22196) and check the value afterwards.